### DISCUSSION TITLE
I2S on STM32F401

### DISCUSSION THREAD

**Discussion by 0x0fe** [stm32duino/Arduino_Core_STM32#2329] [Q&A] [answered]
What would be the best way to use I2S (MEMS mic ICS-41350 for example) on STM32F401? I did nt found anything in the core relate to I2S.

**Comment by 0x0fe** [comment]
for example, how adaptable would the I2S and PDM APIs of the WB be?

https://github.com/GrumpyOldPizza/ArduinoCore-stm32wb/tree/master/libraries/I2S

https://github.com/GrumpyOldPizza/ArduinoCore-stm32wb/blob/master/libraries/PDM/src/PDM.cpp

Or the L4 I2S

https://github.com/GrumpyOldPizza/arduino-STM32L4/blob/master/libraries/I2S/src/I2S.cpp

**Reply by fpistm** [reply]
Not the same core so don't know.

**Reply by 0x0fe** [reply]
yes, it is not the same core, and another one i found :

https://github.com/danieleff/STM32GENERIC/tree/8d5f6722b2bc8ce96943c04473a6d2b6904bbf40/STM32/libraries/I2S

Its also based on a different core, but it relies on the HAL APIs, however in this official core, do we have access to the HAL APIs? i only saw the LL files.

**Reply by 0x0fe** [reply]
ah, the Swiss guy wrote one 

https://github.com/pschatzmann/stm32f411-i2s

How can we include the pdm library : [libPDMFilter_GCC.a](https://github.com/shrayanth-zz/STM32F4/blob/c38cda5475dc24f06106fa563596dea97006836e/STM32F4_Discovery_IO_Toggle/Utilities/STM32F4-Discovery/libPDMFilter_GCC.a)  in the current official core?

https://github.com/shrayanth-zz/STM32F4/tree/c38cda5475dc24f06106fa563596dea97006836e/STM32F4_Discovery_IO_Toggle/Utilities/STM32F4-Discovery

**Reply by fpistm** [reply]
Well, you can overrides somes recipes from platform.txt and boards.txt using the *.local.txt feature. This will allow to give the information needed to link against your library.
https://arduino.github.io/arduino-cli/0.34/platform-specification/#platformlocaltxt
https://arduino.github.io/arduino-cli/0.34/platform-specification/#boardslocaltxt

Or create an arduino libray to give the precompiled property:
https://arduino.github.io/arduino-cli/0.34/library-specification/

But I won't go any further because this goes beyond the scope of basic and purely Arduino use. You can try the forum (stm32duino.com or Arduino).

**Reply by 0x0fe** [reply]
i see, the latter option seems convenient, i will give it a try.
I think it would make sense to support I2S and PDM in the Arduino core, they are often used for various applications such as audio, machine learning / detection / predictive maintenance.

**Reply by fpistm** [reply]
It is a community project, feel free to contribute 😉 

Note that some library used the I2S: ex: https://github.com/stm32duino/X-NUCLEO-IKS02A1-Audio

**Comment by 0x0fe** [comment]
ohh, there is both I2S and PDM here, that looks good

**Comment by 0x0fe** [comment]
i wonder why pdm2pcm(uint8_t *data_in, int16_t *data_out, int16_t size) takes an uint8_t as input buffer, while PDMClass::Record(uint16_t *pBuf) saves the PDM datas to an unint16_t buffer, are we supposed to cast the uint16_t buffer to uint8_t when passing to pdm2pcm?

**Reply by fpistm** [reply]
I don't know @0x0fe.
Maybe @cparata have the answer.

**Comment by cparata** [comment]
Hello @0x0fe ,
PDM and PCM are very different. The above API takes as input a PDM stream data, that is the typical output of a microphone, and transforms it in a PCM data where each PCM sample has a 16 bits size with sign. This is the reason why the output buffer is an array of int16_t.
You can give a look at [this article](https://dsp.stackexchange.com/questions/86862/what-are-the-step-by-step-actions-taken-in-order-to-convert-pdm-data-to-pcm-data) to understand better how the conversion from PDM to PCM is done. Consider that the output PCM stream has a frequency of 16 KHz and a sample size of 16 bits. The PDM frequency in the input stream instead is 1280 KHz.
Best Regards,
Carlo

**Reply by 0x0fe** [reply]
@cparata  i know the difference between PDM and PCM, thank's, my question was : why do you use an uint16_t buffer to store the PDM datas from  PDMClass::Record but then provide an uint8_t format as input for pdm2pcm? it seems like either the buffer of PDMClass::Record should be uint8_t or we should cast the uint16_t buffer to uint8_t when feeding it to pdm2pcm.

**Reply by cparata** [reply]
Hello @0x0fe ,
PDM stream is a bit stream and it does not make sense to manage it with 16 bits samples. Besides, the library that I used to implement the API worked with uint8_t data stream and so I kept the same format as input data.
Best Regards,
Carlo

**Reply by 0x0fe** [reply]
Hello @cparata 
I am a little confused by your answer, everything on the acquisition side is aligned on half words :    
```
hdma_spi3_rx.Init.PeriphDataAlignment = DMA_PDATAALIGN_HALFWORD;
hdma_spi3_rx.Init.MemDataAlignment = DMA_MDATAALIGN_HALFWORD;
int Record(uint16_t *pBuf);
```
So could you provide an example on how you intend your pdm2pm functions to be used along with Record(uint16_t *pBuf);?

```
  uint16_t pdm_bfr[2048];
  int16_t pcm_bfr[2048];

  pdm_begin();
  pdm_record(pdm_bfr);
  delay(10);
  pdm_stop();

  pdm2pcm_init(BYTE_LEFT_MSB,PDM_ENDIANNESS_BE,SINC3);
  pdm2pcm_volume(volume);
  pdm2pcm((uint8_t*)pdm_bfr, pcm_bfr, BLOCK_SIZE);
```
Thank's

**Reply by cparata** [reply, marked as answer]
Hello @0x0fe ,
I meant the implementation of the pdm2pcm API. If you give a look at the LuT_Filter function, it takes as input an array of uint8_t. That's the reason why the first parameter is an array of uint8_t. Indeed, you can give a look at [this example](https://github.com/stm32duino/X-NUCLEO-IKS02A1/blob/main/examples/X_NUCLEO_IKS02A1_Python_Wave_Serial_Encoder/X_NUCLEO_IKS02A1_Python_Wave_Serial_Encoder.ino) based on that library. In this example the input buffer is defined as an array of uint16_t and then it is casted to uint8_t pointer as you mentioned. You can take it as reference.
Best Regards,
Carlo

**Reply by 0x0fe** [reply]
I see, thank you for the reference.

**Comment by 0x0fe** [comment]
i get an linking error with the dsp library : 

```
"C:\\Users\\xxxxxx\\AppData\\Local\\Temp\\arduino\\cores\\d0ea0549b0b1d161b0e0acbc8d76739c\\core.a" -lc -Wl,--end-group -lm -lgcc -lstdc++
c:/users/xxxxxx/appdata/local/arduino15/packages/stmicroelectronics/tools/xpack-arm-none-eabi-gcc/12.2.1-1.2/bin/../lib/gcc/arm-none-eabi/12.2.1/../../../../arm-none-eabi/bin/ld.exe: C:\Users\xxxxxx\AppData\Local\Temp\arduino\sketches\CBCB91AB3DDD889ABB6EF93AB71274DF\sketch\xxxxxxxxxxxxxx.ino.cpp.o: in function `pdm2pcm_volume':
xxxxxxxxxxxxxx.ino.cpp:(.text.pdm2pcm_volume+0x48): undefined reference to `arm_fir_decimate_init_q15'
```

But  arm_math.h is well included, and i could open the header with no issue. It looks like a compiler option or preprocessor flag is missing?
![image](https://github.com/stm32duino/Arduino_Core_STM32/assets/89607149/c9b1384f-1545-47c7-9a11-cffda7a5ac52)

EDIT : Solved , i forgot to include #include "CMSIS_DSP.h"
